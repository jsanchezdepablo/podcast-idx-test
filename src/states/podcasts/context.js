import React, { createContext, useContext, useReducer, useMemo } from "react";
import Reducer from "./reducer";
import INIT_STATE from "./init-state";

const Context = createContext();
export const Provider = ({ children }) => {
  const [state, dispatch] = useReducer(Reducer, INIT_STATE);
  const contextValue = useMemo(() => ({ state, dispatch }), [state, dispatch]);
  return <Context.Provider value={contextValue}>{children}</Context.Provider>;
};
export const Selector = () => useContext(Context);
