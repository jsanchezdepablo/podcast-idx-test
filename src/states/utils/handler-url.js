const PROXY_URL = "https://cors-anywhere.herokuapp.com/";

export const handlerUrl = (url, params) => {
  let processedUrl = PROXY_URL + url;

  if (params !== null) {
    let queryString = Object.keys(params)
      .map((key) => `${key}=${params[key]}`)
      .join("&");

    processedUrl = `${processedUrl}?${queryString}`;
  }

  return processedUrl;
};
