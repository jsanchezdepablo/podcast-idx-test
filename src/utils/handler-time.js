import moment from "moment/moment";

export const getDateParser = (dateUTZ) => moment(dateUTZ).format("DD/MM/YYYY");
export const getMinutesFromMs = (ms) => moment.duration(ms).asMinutes();
export const getHoursFromMs = (ms) => moment.duration(ms).asHours();
export const getIsSpent24hours = (date) => !date || getHoursFromMs(Date.now() - date) >= 24;
