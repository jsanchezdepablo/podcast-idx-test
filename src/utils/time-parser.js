import moment from "moment/moment";

export const getDateParser = (dateUTZ) => moment(dateUTZ).format("DD/MM/YYYY");

export const getMinutes = (ms) => moment.duration(ms).asMinutes();
