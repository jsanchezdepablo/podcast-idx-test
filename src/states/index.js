import React from "react";
import { PodcastsProvider } from "./podcasts";

const GlobalStateProvider = ({ children }) => <PodcastsProvider>{children}</PodcastsProvider>;

export default GlobalStateProvider;
