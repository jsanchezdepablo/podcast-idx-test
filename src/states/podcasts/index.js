import { Provider, Selector } from "./context";
import Actions from "./actions";

export const PodcastsProvider = Provider;
export const usePodcastsSelector = Selector;
export const PodcastsActions = Actions;
