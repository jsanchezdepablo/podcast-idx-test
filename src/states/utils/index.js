import { useState } from "react";
import { handlerUrl } from "./handler-url";

export const useFetchApi = (url) => {
  const [result, setResult] = useState(null);
  const [isLoading, setIsLoading] = useState(false);
  const [lastUpdated, setLastUpdated] = useState();
  const [status, setStatus] = useState();

  const search = (params = null) => {
    setIsLoading(true);
    setResult(null);
    setStatus();

    return fetch(handlerUrl(url, params))
      .then((response) => {
        setStatus(response.status);
        setLastUpdated(new Date().getTime());
        return response.json();
      })
      .then((result) => {
        setIsLoading(false);
        setResult(result);
      })
      .catch((e) => {
        setStatus(e.status);
        console.error(e);
      })
      .finally(() => {
        setIsLoading(false);
      });
  };

  return { search, result, isLoading, lastUpdated, status };
};
