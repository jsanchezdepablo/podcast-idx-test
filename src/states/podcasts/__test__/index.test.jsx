import React from "react";
import { shallow } from "enzyme";
import { PodcastsProvider, PodcastsActions } from "..";
import Reducer from "../reducer";
import ACTION_TYPES from "../action-types";
import INIT_STATE from "../init-state";
import * as Context from "../context";
import * as Utils from "states/utils";
import * as HandlerTimeFunctions from "utils/handler-time";

jest.mock("react", () => {
  return {
    ...jest.requireActual("react"),
    useContext: (f) => f,
  };
});

describe("Podcasts state", () => {
  beforeEach(() => {
    jest.restoreAllMocks();
    jest.resetAllMocks();
  });

  it("Podcasts Provider", () => {
    const wrapper = shallow(
      <PodcastsProvider>
        <div className="test-class" />
      </PodcastsProvider>,
    );
    expect(wrapper.find(".test-class").exists()).toBe(true);
    expect(wrapper).toMatchSnapshot();
  });

  it("Podcasts Actions - usePodcastActions", () => {
    const mockDispatch = jest.fn();
    jest.spyOn(Context, "Selector").mockImplementation(() => ({ dispatch: mockDispatch }));

    let resultUsePodcastActions = null;
    const UsePodcastActions = () => {
      resultUsePodcastActions = PodcastsActions.usePodcastActions();
      return <div />;
    };

    shallow(<UsePodcastActions />);
    expect(mockDispatch).toHaveBeenCalledTimes(0);

    resultUsePodcastActions.setIsFirstStart();
    expect(mockDispatch).toHaveBeenCalledTimes(1);
  });

  it("Podcasts Actions - useFetchPodcasts", () => {
    const mockDispatch = jest.fn();
    const mockSearch = jest.fn();
    const useFetchOneApi = jest.spyOn(Utils, "useFetchApi");

    useFetchOneApi.mockImplementation(() => ({ result: [], isLoading: false, search: mockSearch }));

    jest.spyOn(Context, "Selector").mockImplementation(() => ({ dispatch: mockDispatch }));
    jest.spyOn(HandlerTimeFunctions, "getIsSpent24hours").mockImplementation(() => true);

    let resultUseFetchPodcasts = null;
    const UseFetchPodcasts = () => {
      resultUseFetchPodcasts = PodcastsActions.useFetchPodcasts();
      return <div />;
    };

    shallow(<UseFetchPodcasts />);
    expect(mockSearch).toHaveBeenCalledTimes(0);

    const args = { date: "date" };
    resultUseFetchPodcasts.searchPodcasts(args.date);
    expect(mockSearch).toHaveBeenCalledTimes(1);
  });

  it("Podcasts Actions - useFetchEpisodes", () => {
    const mockDispatch = jest.fn();
    const mockSearch = jest.fn();
    const useFetchOneApi = jest.spyOn(Utils, "useFetchApi");

    useFetchOneApi.mockImplementation(() => ({ result: [], isLoading: false, search: mockSearch }));

    jest.spyOn(Context, "Selector").mockImplementation(() => ({ dispatch: mockDispatch }));
    jest.spyOn(HandlerTimeFunctions, "getIsSpent24hours").mockImplementation(() => true);

    let resultUseFetchEpisodes = null;
    const UseFetchEpisodes = () => {
      resultUseFetchEpisodes = PodcastsActions.useFetchEpisodes();
      return <div />;
    };

    shallow(<UseFetchEpisodes />);
    expect(mockSearch).toHaveBeenCalledTimes(0);

    const args = { id: "id", date: "date" };
    resultUseFetchEpisodes.searchEpisodes(args.id, args.date);
    expect(mockSearch).toHaveBeenCalledTimes(1);
  });

  it("Podcasts Reducer", () => {
    const payload = { name: "test" };
    expect(Reducer({}, { type: "test" })).toEqual({});
    expect(Reducer({}, { type: ACTION_TYPES.SET_IS_LOADING, payload })).toMatchObject({ isLoading: payload });
    expect(Reducer({}, { type: ACTION_TYPES.SET_IS_FIRST_START, payload })).toMatchObject({ isFirstStart: payload });
    expect(Reducer({}, { type: ACTION_TYPES.SET_LAST_PODCASTS_UPDATED_DATE, payload })).toMatchObject({
      lastUpdatedDate: payload,
    });
    expect(Reducer({}, { type: ACTION_TYPES.RESET })).toEqual(INIT_STATE);
  });
});
