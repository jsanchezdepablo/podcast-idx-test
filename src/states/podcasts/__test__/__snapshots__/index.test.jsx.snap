// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Podcasts state Podcasts Provider 1`] = `
<ContextProvider
  value={
    Object {
      "dispatch": [Function],
      "state": Object {
        "isFirstStart": true,
        "isLoading": false,
        "lastUpdatedDate": null,
        "podcasts": Array [],
      },
    }
  }
>
  <div
    className="test-class"
  />
</ContextProvider>
`;
